### PR TITLE
Make Additional Function Capture Precise

### LINKED ISSUE #1093: Make additional functions capture precisely the values they need
Currently additional functions serialize all values that their original body captured in case the rewritten body uses those values. In general this is an overapproximation and may lead to emitting some otherwise unreachable code.

e.g.:
```
function Bar() {
  return 123;
}

function Foo() {
  return Bar();
}

__registerAdditionalFunctionToPrepack(Foo);

global.Foo = Foo;
```

results in
```
(function () {
  "use strict";

  var _$0 = this;

  var _0 = function () { // Foo
    return 123;
  };

  var _1 = function () { // Bar
    return 123;
  };

  _$0.Foo = _0;
}).call(this);
```

Make the visiting/emission of values captured by additional functions precise by not visiting them all upfront and instead making sure that captured values that are used will be visited for some other reason. The easiest solution to this is probably making all captured values AbstractValues during evaluation of the additional function so that any uses will show up in the generator and only captured values that are operated on in the final code (those that cause generator entries) will be visited.

### PR DESCRIPTION
**Pull request by cblappert**
Release Notes: none.

This fixes #1093. At a high level, captures are currently needed by the additional function only if they are modified.

Implementation-wise we tentatively create all bindings with callbacks to visit them and delete them to make sure that after Additional Functions are visited we have a precise set of captures.

Functionality of the PR is fully there but I would like to refactor it in some way because I feel like it can be cleaner.

### CODE REVIEW

File: src/serializer/ResidualHeapVisitor.js [hunk 1]
```diff
@@ -525,6 +583,8 @@ export class ResidualHeapVisitor {
       binding = ({ value, modified: true, declarativeEnvironmentRecord: null }: ResidualFunctionBinding);
       this.globalBindings.set(key, binding);
     }
+    let oldValue = binding.value;
+    invariant(!oldValue || oldValue instanceof Value);
```
Comment (NTillmann) [inline src/serializer/ResidualHeapVisitor.js]: Is this invariant needed for Flow?

File: src/serializer/ResidualHeapVisitor.js [hunk 2]
```diff
@@ -351,6 +400,10 @@ export class ResidualHeapVisitor {
       for (let innerName of functionInfo.unbound) {
         let residualFunctionBinding;
         let doesNotMatter = true;
+        let modified = functionInfo.modified.has(innerName);
+        // For additonal function values, we don't know if non-modified values
```
Comment (NTillmann) [inline src/serializer/ResidualHeapVisitor.js]: additional

Comment (NTillmann) [reply src/serializer/ResidualHeapVisitor.js]: Why don't we know that? Are we lacking a visit functions for the `Effects` that define additional functions?

**Comment by cblappert** [comment]
created #1140 to address comments

**Pull request closed without merging**